### PR TITLE
Created Category model

### PR DESCRIPTION
**Pull request by vikkushnir**

### CODE REVIEW

File: src/main/java/online/book/store/mapper/CategoryMapper.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import online.book.store.model.Category;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface CategoryMapper {
+    CategoryResponseDto toDto(Category category);
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "deleted", ignore = true)
```
Comment (MolchanovAlexander) [inline src/main/java/online/book/store/mapper/CategoryMapper.java]: I have none of this in my CategoryMapper and it works.

File: src/main/java/online/book/store/mapper/BookMapper.java [hunk 2]
```diff
@@ -1,17 +1,39 @@
...
 import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
     BookResponseDto toDto(Book book);
 
     @Mapping(target = "id", ignore = true)
     @Mapping(target = "deleted", ignore = true)
+    @Mapping(target = "categories", ignore = true)
```
Comment (MolchanovAlexander) [inline src/main/java/online/book/store/mapper/BookMapper.java:21]: "Mapping(target = "categories", ignore = true)" - I have only this one .

**Review by MolchanovAlexander — APPROVED** [review]
Not bad!

File: src/main/java/online/book/store/mapper/CategoryMapper.java (hunk 1 above)
Comment (vikkushnir) [reply src/main/java/online/book/store/mapper/CategoryMapper.java]: It also works on my side, but, I had a warn message in the console, and after adding these, the message was gone, but now, when I deleted it, that message doesn't come back, weird

File: src/main/java/online/book/store/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Get books by category ID",
+            description = "Get books by category ID")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id,
+                                                                Pageable pageable) {
+        return bookService.getBooksByCategoryId(id, pageable);
+    }
+
```
Comment (andrii-hoienko) [inline src/main/java/online/book/store/controller/CategoryController.java]:
```suggestion
```

File: src/main/java/online/book/store/dto/request/category/CategoryRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+
+@Data
+public class CategoryRequestDto {
+    @NotBlank
+    @Size(max = 255)
+    private String name;
+    @Size(max = 255)
+    private String description;
+
```
Comment (andrii-hoienko) [inline src/main/java/online/book/store/dto/request/category/CategoryRequestDto.java]:
```suggestion
```

File: src/main/java/online/book/store/dto/request/category/CategoryRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class CategoryRequestDto {
+    @NotBlank
+    @Size(max = 255)
+    private String name;
+    @Size(max = 255)
+    private String description;
```
Comment (andrii-hoienko) [inline src/main/java/online/book/store/dto/request/category/CategoryRequestDto.java:14]: Could description be null?

File: src/main/java/online/book/store/service/category/CategoryServiceImp.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            throw new EntityNotFoundException("Can't find category with id: " + id);
+        }
+        Category category = categoryMapper.toEntity(requestDto);
+        category.setId(id);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        categoryRepository.deleteById(id);
```
Comment (andrii-hoienko) [inline src/main/java/online/book/store/service/category/CategoryServiceImp.java]: Please check if category exists

File: src/main/java/online/book/store/service/book/BookServiceImpl.java [hunk 7]
```diff
@@ -28,14 +29,14 @@ public BookResponseDto save(BookRequestDto requestDto) {
     }
 
     @Override
-    public List<BookResponseDto> findAll(Pageable pageable) {
+    public List<BookResponseDto> getAll(Pageable pageable) {
```
Comment (andrii-hoienko) [inline src/main/java/online/book/store/service/book/BookServiceImpl.java:51]: You should also modify save method to ensure that categories that you are trying to add to book exist

File: src/main/java/online/book/store/mapper/CategoryMapper.java (hunk 1 above)
Comment (andrii-hoienko) [reply src/main/java/online/book/store/mapper/CategoryMapper.java]: Then please remove unnecessary annotations

File: src/main/java/online/book/store/mapper/BookMapper.java (hunk 2 above)
Comment (andrii-hoienko) [reply src/main/java/online/book/store/mapper/BookMapper.java:21]: Investigate is it possible to remove unnecessary annotations

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Good job! Left some comments

File: src/main/java/online/book/store/service/book/BookServiceImpl.java [hunk 8]
```diff
@@ -20,22 +25,31 @@ public class BookServiceImpl implements BookService {
...
     public BookResponseDto save(BookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
+        Set<Category> categories = requestDto.getCategoryIds()
+                .stream()
+                .map(categoryId
+                        -> categoryRepository.findById(categoryId)
+                        .orElseThrow(() -> new EntityNotFoundException(
+                                "Invalid category ID: " + categoryId)))
+                .collect(Collectors.toSet());
+        book.setCategories(categories);
```
Comment (andrii-hoienko) [inline src/main/java/online/book/store/service/book/BookServiceImpl.java]: In this implementation you make N requests to db for each category, it is not efficient, better to take all existing categories dy id`s that comes with BookRequestDto  and then check size of obtained list

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left one comment

**Pull request merged**